### PR TITLE
Fix testing/ci to run probes against Firefox

### PR DESCRIPTION
**Pull request by tdeekens**
> This pull request changes that we run PRs against one browser on CI for PRs.

### Background & Context

This adds behaviour to the karma run so that we run against a randomly selected browser on PRs. This helps not to merge failing PRs as they only run their tests against JSDOM. It also simplifies the complications in the package.json of reading our `TEST_BROWSERSTACK` and Travis' `TRAVIS_PULL_REQUEST` env variable. 

- `TEST_BROWSERSTACK` is passed in as `shouldTestBrowserstack`
- `TRAVIS_PULL_REQUEST` as `shouldProbeOnly`

Those env vars are then read in the karma.conf.js to either

- Select a random browser to run against BrowserStack ('shouldProbeOnly')
- Allow to specify a custom browser e.g. `--browsers Chrome` for local development

I am currently out of my limit on BrowserStack so I'll wait for a few minutes. After I'll investigate why actually CI tests only fail for a few browsers currently (as it seems).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "build:cjs": "cross-env NODE_ENV=development BABEL_ENV=rollup rollup -c -f cjs -o dist/purify.cjs.js",
     "test:jsdom": "cross-env NODE_ENV=test BABEL_ENV=rollup node test/jsdom-node-runner --dot",
     "test:karma": "cross-env NODE_ENV=test BABEL_ENV=rollup karma start test/karma.conf.js --log-level warn ",
-    "test:ci": "cross-env NODE_ENV=test BABEL_ENV=rollup npm run lint && npm run test:jsdom && (([ \"${TRAVIS_PULL_REQUEST}\" != \"false\" ] || [ \"${TEST_BROWSERSTACK}\" != \"true\" ]) || karma start test/karma.conf.js --log-level error --reporters dots --single-run)",
+    "test:ci": "cross-env NODE_ENV=test BABEL_ENV=rollup npm run test:karma -- --shouldTestBrowserstack=\"${TEST_BROWSERSTACK}\" --shouldProbeOnly=\"${TRAVIS_PULL_REQUEST}\"",
```
Comment (tdeekens) [inline package.json]: Mapping of things.

File: package.json [hunk 2]
```diff
@@ -41,6 +41,12 @@
           "trailingComma": "es5",
           "singleQuote": true
         }
+      ],
+      "camelcase": [
```
Comment (tdeekens) [inline package.json]: ...so linting does not break on karma config

File: test/karma.conf.js [hunk 3]
```diff
@@ -1,6 +1,19 @@
...
+const {
+  customLaunchers,
+  browsers,
+  browser,
+} = require('./karma.custom-launchers.config');
+
+const shouldProbeOnly = argv.shouldProbeOnly === 'true';
+const shouldTestBrowserstack = argv.shouldTestBrowserstack === 'true';
+const browserStackBrowsers = shouldProbeOnly ? [browser] : browsers;
+const karmaBrowsers = shouldTestBrowserstack
```
Comment (tdeekens) [inline test/karma.conf.js]: Switch on things

File: test/karma.custom-launchers.config.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
+const sample = require('lodash.sample');
+
+const customLaunchers = {
```
Comment (tdeekens) [inline test/karma.custom-launchers.config.js]: ...all custom BS browsers

File: test/karma.custom-launchers.config.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    os_version: '10',
+  },
+};
+
+const getAllBrowsers = () => Object.keys(customLaunchers);
+const getRandomBrowser = () => sample(getAllBrowsers());
+
+module.exports = {
+  customLaunchers,
+  browsers: getAllBrowsers(),
```
Comment (tdeekens) [inline test/karma.custom-launchers.config.js]: Export the launchers and the browsers

**Comment by cure53** [comment]
Hm, nice!

**Comment by tdeekens** [comment]
Sadly we can not run on BrowserStack if the PR comes from anywhere except origin so any fork 😢  https://docs.travis-ci.com/user/pull-requests/

File: .travis.yml [hunk 6]
```diff
@@ -2,8 +2,14 @@ sudo: false
 language: node_js
 cache: yarn
 script: npm run test:ci
+dist: trusty
```
Comment (tdeekens) [inline .travis.yml]: When using the newer trusty images on TravisCI (https://docs.travis-ci.com/user/reference/trusty/) we can run on the latest Firefox on PRs at least.

**Comment by tdeekens** [comment]
This all would mean testing would be

PR:
node 4 (JSDOM), 5 (JSDOM), 6  (JSDOM), 8 (JSDOM, Firefox latest) 

master:
node 4 (JSDOM), 5 (JSDOM), 6  (JSDOM), 8 (JSDOM, BrowserStack fleet)

**Comment by tdeekens** [comment]
Currently locally on my machine running against all browsers on BrowserStack hoping my free plan to allow for it. So far everything is green. I wonder where the error was/is.

**Comment by tdeekens** [comment]
screen shot 2017-08-20 at 22 47 41

All green for me with all browsers against BrowserStack from my local machine. Somebody else wants to give it a try?

**Comment by cure53** [comment]
I can, ready to merge?

**Comment by tdeekens** [comment]
Give me a sec I am on the heels to the cause of the whole mess https://github.com/vimeo/rollup-plugin-bundle-size/issues/2

**Comment by cure53** [comment]
Aye :)

**Comment by tdeekens** [comment]
Madness. So rollup (our bundler) bumped to 0.48 with some configuration changes but none really breaking and being graceful in fallbacks. However the semver matching for non 1.0.0 releases is a big different which lead to our `karma-rollup-preprocessor` to get a newer version in its node_modules (why, I don't know. Dependency resolution things). Which melted into one rollup not understanding our `rollup.config.js` and the other doing. No matter how you flipped the coin, you couldn't satisfy both. At some point I managed to both resolve to `0.48` but then Karma would not boot failing with "No Captured Browser".

The current `yarn.lock` and `package-lock.json` should pin things and I removed the bundle-size-plugin for now. So our build and test setup problem is resolved, we test Firefox on PRs but some day need to fix this rollup versioning clash.

**Comment by tdeekens** [comment]
Sadly this all erupted in a 1.0.0 release and I am really sorry for that!

**Comment by cure53** [comment]
Wow. Okay, let's give it  a try!

**Comment by cure53** [comment]
Works well on FF indeed!

**Comment by tdeekens** [comment]
Cool. But wait a minute. I think we didn't run against all BrowserStack browsers on `master`. Let me try to fix that quickly.

**Comment by cure53** [comment]
Yeah, just FF for now.

**Comment by tdeekens** [comment]
I think it's fine if we run FF on PRs on all node versions? It might make the chain a bit longer but avoid a bit strange complexity in the build matrix.

**Comment by cure53** [comment]
I think so, yes.

**Pull request merged**